base/acme/src/main/java/org/dogtagpki/acme/database/LDAPDatabase.java

```diff
@@ -0,0 +1,650 @@
+            Vector<ACMEChallenge> challenges = new Vector<>();
+            Collection<LDAPEntry> entries = ldapSearch(
+                RDN_CHALLENGE + "," + basedn,
+                "(&(" + ATTR_OBJECTCLASS + "=" + OBJ_CHALLENGE +
+                    ")(" + ATTR_AUTHORIZATION_ID + "=" + authzID + "))"
+            );
+            for (LDAPEntry challengeEntry : entries) {
+                challenges.add(loadChallenge(challengeEntry));
+            }
+            authz.setChallenges(challenges);
+        }
+
+        return authz;
+    }
+
+    private ACMEChallenge getChallenge(String challengeID)
+            throws Exception {
+        String dn = ATTR_CHALLENGE_ID + "=" + challengeID
+                        + "," + RDN_CHALLENGE + "," + basedn;
+        LDAPEntry entry = ldapGet(dn);
+        if (entry == null) return null;
+        return loadChallenge(entry);
+    }
+
+    private ACMEChallenge loadChallenge(LDAPEntry entry)
+            throws ParseException {
+        ACMEChallenge challenge = new ACMEChallenge();
+
+        LDAPAttribute attr = entry.getAttribute(ATTR_CHALLENGE_ID);
+        challenge.setID(attr.getStringValues().nextElement());
+
+        attr = entry.getAttribute(ATTR_OBJECTCLASS);
+        List<String> classes = Collections.list(attr.getStringValues());
+        if (classes.contains(OBJ_CHALLENGE_DNS01)) {
+            challenge.setType("dns-01");
+        } else if (classes.contains(OBJ_CHALLENGE_HTTP01)) {
+            challenge.setType("http-01");
+        } else {
+            challenge.setType("unknown-challenge-type");  // shouldn't happen
+        }
+
+        attr = entry.getAttribute(ATTR_AUTHORIZATION_ID);
+        challenge.setAuthzID(attr.getStringValues().nextElement());
+
+        attr = entry.getAttribute(ATTR_TOKEN);
+        challenge.setToken(attr.getStringValues().nextElement());
+
+        attr = entry.getAttribute(ATTR_STATUS);
+        challenge.setStatus(attr.getStringValues().nextElement());
+
+        attr = entry.getAttribute(ATTR_ERROR);
+        if (attr != null) {
+            challenge.setError(attr.getStringValues().nextElement());
+        }
+
+        attr = entry.getAttribute(ATTR_VALIDATED_AT);
+        if (attr != null) {
+            challenge.setValidationTime(
+                dateFormat.parse(attr.getStringValues().nextElement()));
+        }
+
+        return challenge;
+    }
+
+    /** Get the authorization for the given challenge.
+     * NOTE: we do not load all challenges for the authorization,
+     * only the challenge with the given challengeID.
+     */
+    public ACMEAuthorization getAuthorizationByChallenge(String challengeID) throws Exception {
+        ACMEChallenge challenge = getChallenge(challengeID);
+        if (challenge == null) return null;
+
+        ACMEAuthorization authz = getAuthorization(challenge.getAuthzID(), LoadChallenges.DontLoad);
+        if (authz == null) return null;
+
+        authz.setChallenges(Collections.singleton(challenge));
```

Comment:
Got to construct a `Collection<ACMEChallenge>` somehow.  Seems good to me.

Reply:
In case we're writing a unit test that modifies the challenges in an authorization object obtained using this method the test might fail since the list of challenges is immutable. Also if the test tries to verify all challenges associated with the authorization object obtained using this method the test might fail too since the method doesn't load all challenges. It's probably not a problem now, but it could be a problem in the future.

Reply:
Oh I didn't realise it was immutable. 

```
nobody: ...
java: let's return data that violate their own interface contacts!
```

Sigh.  I'll change it.